pkg/actor/system.go

```diff
@@ -68,18 +90,58 @@ func (p *proc) batchReceiveMsgs(batchMsg []message.Message) int {
 	return n
 }
 
-// close its actor.
-// close is threadsafe.
-func (p *proc) close() {
-	atomic.StoreUint64(&p.closed, 1)
-}
-
-// isClosed returns ture, means its actor is closed.
+// isClosed returns ture, means its mailbox and actor are closed.
 // isClosed is threadsafe.
 func (p *proc) isClosed() bool {
-	return atomic.LoadUint64(&p.closed) == 1
+	return atomic.LoadUint64(&p.state) == uint64(procStateClosed)
+}
+
+// closeMailbox close mailbox and set state to closed for graceful close.
+// onSystemStop is threadsafe.
+func (p *proc) onSystemStop() {
+	// Running -> MailboxClosed
+	if atomic.CompareAndSwapUint64(
+		&p.state, uint64(procStateRunning), uint64(procStateMailboxClosed)) {
+		p.mb.close()
+	}
 }
 
+// closeMailbox close mailbox and set state to closed.
+// onMailboxEmpty is threadsafe.
+func (p *proc) onMailboxEmpty() {
+	// MailboxClosed -> Close
+	if atomic.CompareAndSwapUint64(
+		&p.state, uint64(procStateMailboxClosed), uint64(procStateClosed)) {
+		p.actor.Close()
+	}
+}
+
+// onActorClosed all mailbox and actor.
+// onActorClosed is threadsafe.
+func (p *proc) onActorClosed() {
+	if atomic.CompareAndSwapUint64(
+		&p.state, uint64(procStateRunning), uint64(procStateClosed)) {
+		p.mb.close()
+		p.actor.Close()
+		return
+	}
+	if atomic.CompareAndSwapUint64(
+		&p.state, uint64(procStateMailboxClosed), uint64(procStateClosed)) {
+		p.actor.Close()
+	}
+}
+
+type readyState int
+
+const (
+	// Running state, system needs to poll actors.
+	readyStateRunning readyState = 0
```

Comment:
how about `readyStateRunning readyState = itoa`, and then ignore the following enum number.